### PR TITLE
Option to set a validate time when calling the validator

### PR DESCRIPTION
**Pull request by notbrokenjemma**
Option to pass a "now" time to the verifiers, so we can check how a signature would have validated at a given timestamp.

### CODE REVIEW

File: lib/Mail/DKIM/ARC/Verifier.pm [hunk 1]
```diff
@@ -193,6 +193,11 @@ sub add_signature {
 
     return if $self->{result};    # already failed
 
+    # Set verification time if we have one
+    if ($self->{verify_time}) {
+        $signature->{_verify_time} = $self->{verify_time};
```
Comment (Davis-A) [inline lib/Mail/DKIM/ARC/Verifier.pm]: I know it's only convention, but I strongly dislike assignment on a `_private` key, and from a quick look there's nowhere else in the codebase where assignment to a private key is done on anything but `$self`

File: t/signer_expiration_set_time.t [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+use Mail::DKIM::Signer;
+use Mail::DKIM::Verifier;
+
+my $homedir = ( -d "t" ) ? "t" : ".";
+
+my $tdir    = -f "t/test.key" ? "t" : ".";
+my $keyfile = "$tdir/test.key";
+
+sub generate_signed_email {
```
Comment (Davis-A) [inline t/signer_expiration_set_time.t:16]: Is it worth creating a utility module in `t/` so there isn't code repetition?

File: t/signer_expiration_set_time.t [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+$verifier = Mail::DKIM::Verifier->new();
+$verifier->PRINT($signed_email);
+$verifier->CLOSE;
+is( $verifier->result, 'pass', 'Non-expired Signature does pass');
+
+# override the DNS implementation, so that these tests do not
+# rely on DNS servers I have no control over
+my $CACHE;
+
+sub Mail::DKIM::DNS::fake_query {
```
Comment (Davis-A) [inline t/signer_expiration_set_time.t:71]: ditto

**Review by Davis-A — CHANGES_REQUESTED** [review]
If you twist my arm I'll approve it as is, but I think it could do with some neatening

File: lib/Mail/DKIM/ARC/Verifier.pm (hunk 1 above)
Comment (notbrokenjemma) [reply lib/Mail/DKIM/ARC/Verifier.pm]: Have created a setter and used that instead

File: t/signer_expiration_set_time.t (hunk 3 above)
Comment (notbrokenjemma) [reply t/signer_expiration_set_time.t:71]: This should/could all be replaced with passing Net::DNS::Resolver::Mock as the resolver, which would be much tidier and remove fake DNS code from the tests altogether, just pass in a zonefile and we're done.
We should check to make sure there are no Perl version issues, but I'm going to argue that this, and the previous tests cleanup comment should be their own task and MR.

File: lib/Mail/DKIM/ARC/Verifier.pm (hunk 1 above)
Comment (Davis-A) [reply lib/Mail/DKIM/ARC/Verifier.pm]: Cheers.  I know it's a small thing, but I really do think making it clear what data is encapsulated and isn't is a distinction worth upholding

File: t/signer_expiration_set_time.t (hunk 3 above)
Comment (Davis-A) [reply t/signer_expiration_set_time.t:71]: I'll take that as a reasonable answer.  In an ideal world that'd be done first but we don't live in one.

**Review by Davis-A — APPROVED** [review]

**Pull request merged**